### DISCUSSION TITLE
DevOps success for business leaders: Common pitfalls to avoid

### DISCUSSION THREAD

**Discussion by ayodejiayodele** [community/community#41489] [Discover: GitHub Best Practices]
Have you decided to invest in DevOps, or are you looking to consolidate on the investments you’ve made? 
Let’s look at some of the common pitfalls that often plague DevOps teams or organisations intending to embark on a cultural shift, adopting DevOps practices.

### TL;DR
* [Not leveraging the full potential of cloud capabilities](#1-missing-the-point-on-cloud)
* [Making wholesale changes - going all in](#2-making-wholesale-changes)
* [Measuring the wrong metric](#3-measuring-the-wrong-metric) 
* [Lack of support from leadership](#4-lack-of-support-from-leadership) 
* [Ignoring other capabilities (DevOps is not just about CI/CD)](#5-devops--cicd-)
* [Fear of failure (DevOps is all about CI/CD)](#6-not-failing)
* [Leaving it to the development teams](#7-leaving-it-to-the-development-teams)

## Some common pitfalls

<a id="1-missing-the-point-on-cloud"></a>
### 1. Missing the point on cloud

How cloud are you? - That’s a question I ask teams when they say they are using cloud, or they have moved to the cloud.

Cloud computing, one of the 40+ practices/capabilities of DevOps, is a vital capability that many teams underestimate. Many businesses know the efficiency and value that cloud computing brings, but do not utilise this advantage to its fullest. [Accelerate](https://www.amazon.com.au/Accelerate-Building-Performing-Technology-Organizations/dp/B07BMB9YY2) - a book I highly recommend for teams looking to become high performing tech orgs - re-iterates the meaning of cloud computing as defined by the [NIST](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-145.pdf). If you say you are in the cloud, then all of the following must be true for your applications and systems:
1. On-demand self-service - I can consume or get requests fulfilled from the application/service unassisted.
2. Broad network access - It is available over the network in as much as your business and security rules permit.
3. Supports resource pooling - It sustainably shares or reuses resources based on demand to avoid waste.
4. Rapid elasticity - Your application/service can provision capabilities to scale rapidly to meet surge in demand, and release when it contracts.
5. Measured service - There’s transparency in the metering of usage and the customer has some control. You can also automatically control and optimise from your end.

<a id="2-making-wholesale-changes"></a>
### 2. Making wholesale changes

Due to the huge cost savings and efficiency that can be gained, many leaders are tempted to make wholesale changes too early in the adoption process. As [mentioned before](https://github.com/orgs/community/discussions/33143), DevOps is a continuous experimental journey, and change takes time. You don’t want to lose steam halfway through the journey. You also need to consolidate the gains. Most successful and impactful DevOps transformation programs (e.g. [Microsoft](https://azure.microsoft.com/en-us/resources/devops-at-microsoft-driving-culture-change/), [Target](https://www.youtube.com/watch?v=1FMktLCYukQb), [Capital One](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwj7xaPIzsH7AhV5-TgGHdqmDu0QFnoECBkQAw&url=https%3A%2F%2Fwww.simform.com%2Fblog%2Fcapital-one-devops-case-study%2F%23%3A~%3Atext%3DThe%2520DevOps%2520journey%2520of%2520Capital%2520One%253A%2520the%2520when%2520and%2520the%2520why%2C-In%25202010%252C%2520Capital%26text%3DCapital%2520One%2520focused%2520on%2520hiring%2Cfoundation%2520of%2520the%2520DevOps%2520implementation.&usg=AOvVaw15GfZ7CgbmEFDyB3TmKOG0), [Mercedes-Benz](https://github.com/customer-stories/mercedes-benz)) were carried out in small batches, one change at a time, few teams at a time. 

<img width="600" alt="DevOps 1000 miles" src="https://user-images.githubusercontent.com/63050478/204669043-c398bbc5-d6ae-4c5c-aeab-be79e590d133.png">

<a id="3-measuring-the-wrong-metric"></a>
### 3. Measuring the wrong metric

Are you in the bubble of measuring vanity metrics? Businesses often measure the wrong metrics or measure the right metrics wrongfully. As summarised by [Dominica Degrandis](https://ddegrandis.com/#:~:text=About%20me) (a protagonist of flow methods and making work visible), two important questions that should help identify if a metric is good for your business are:

- Does it matter to my customer?
- Does it help us make a decision that makes our business fitter?

For example, say you are measuring performance for Dev by the total number of lines of code written, or Ops by the total number of tickets assigned. What behaviour(s) are you encouraging? Will it encourage them to deliberately increase the number of lines they write, or choose low-hanging tickets so they can close more in the same timeframe? How does this help the customer or add value to the business?
>
> “Tell me how you will measure me and I will tell you how I’ll behave”
> 

The metrics with which you measure may push teams to game the system. Furthermore, it might also serve you to not only keep an eye on a single metric but rather, a holistic view along with other metrics.

🎥 [Watch this video](https://itrevolution.com/articles/dominica-degrandis-data-playlist/#:~:text=Having%20a%20balanced%20set%20of%20good%20metrics) (20 minutes) of a presentation by Julia Wester and Troy Magennis at the _DevOps Enterprise Summit in 2015_, it really resonated with me. 
 
<a id="4-lack-of-support-from-leadership"></a>
### 4. Lack of support from leadership 

Transformational Leadership has a powerful impact on results. How you inspire your team can affect productivity, efficiency, customer satisfaction, and even profit. DevOps transformation programs can fail if the leadership does not play a role or only ~~pays lip service to~~ sweet-talks the mission. Leadership is important to:

1. Remove silos across teams 
2. Enable and reward the culture of experimentation and innovation 
3. Create a high trust [Westrum organisation (Generative)](https://cloud.google.com/architecture/devops/devops-culture-westrum-organizational-culture)
4. Create a high-level business and technology strategy 

Lacking these can significantly increase the risk of failure or wasted investment.

<a id="5-devops--cicd-"></a>
### 5. DevOps == CI/CD ?

DevOps is not just about CI/CD. Many teams tend to focus on the Continuous Integration (CI) and the Continuous Delivery/Deployment (CD) capabilities, but then neglect Continuous Monitoring (Observability), Continuous Feedback, Continuous Planning, Continuous Security (the security hat), and other capabilities and practices that ought to be injected in the value stream. You then end up with products that continually release new features, but are extremely poor in managing bugs, mitigating and recovering from security incidents and downtimes, supporting and listening to customers, architectural design and scalability, and bridging inter-team silos.

<a id="6-not-failing"></a>
### 6. Not failing

I’ve heard many teams say, _we are doing DevOps but we are not yet releasing to the production environment_. DevOps cannot succeed if you don’t prepare to fail. The fear of breaking something in production has pushed some businesses to add more red tape in-between and elongate the lead time, or even exclude automated deployments to production entirely. Still, such businesses are not immune to the failures they tried to prevent. On top of that, they are not well-prepared to recover from them because they don’t deploy often and are less used to the rigours.

<img width="400" alt="The more you fail, the more you are likely to succeed" src="https://user-images.githubusercontent.com/63050478/204671415-3dd815c5-c883-478b-b34e-455dc59c9005.png">

DevOps is all about CI/CD 🤔. I know this sounds contradictory to the previous point, but it is not. What I mean here is that CI and CD, together, are the bedrock of DevOps. Without them, you cannot really succeed. It is upon these two that many practices rest. You `must` implement CI/CD, but not only CI/CD. With CI/CD, you can implement _[intelligent failure](https://hbr.org/2011/04/failing-by-design#:~:text=An%20alternative%20to%20ignoring%20failure%20is%20to%20foster%20%E2%80%9Cintelligent%20failure%2C%E2%80%9D)_, doing it more often until it becomes less painful.

<a id="7-leaving-it-to-the-development-teams"></a>
### 7. Leaving it to the development teams

For DevOps to work, you need both the Dev and the Ops to work together. Historically, most DevOps initiatives begin with the development teams. Therefore over time, many businesses get carried away and forget to be inclusive. Operations functions such as infrastructure management, incident management, support, performance (system) management, and in some cases, finance and procurement, must all be part of the DevOps conversation, and all need some level of automation/handshake for DevOps to be successful.

I hope this post sheds some light on possible pitfalls to avoid in your DevOps journey. I have also provided some links below for additional reading that may be helpful.

### Useful links
1. [How they did it, 6 companies that scaled DevOps](https://techbeacon.com/app-dev-testing/how-they-did-it-6-companies-scaled-devops) - Article on TechBeacon
2. [Measure what matters](https://www.forbes.com/sites/joelshapiro/2022/01/31/measure-what-matters-how-to-choose-the-right-metrics/?sh=37f798a623c3) - Forbes
3. [Failing by design](https://hbr.org/2011/04/failing-by-design) - Harvard Business Review 
4. [Learning Through Failure: The Strategy of Small Losses](https://scholars.duke.edu/display/pub913886) - Sitkin, Sim B.
5. [How to improve digital and self-service effectiveness](https://www.gartner.com/document/4020612?ref=ki-21257) - Gartner

**Comment by je395191059** [comment]
新手小白，希望能在这里学到知识

**Reply by KittyChiu** [reply]
在这篇文章中，你有什么特别感兴趣想学?

**Comment by michealvaughan** [comment]
Yeah, this conversation is good. Don't forget to check my website https://apkminion.com.

**Comment by michealvaughan** [comment]
This [article](https://apkminion.com/uber-mod-apk-6/) is making issues please help.

**Comment by ghost** [comment]
Plis kembalikan data saya yang lama

**Comment by AngelofVerdant** [comment]
This is a little bit of an eye-opener, I am a beginner and I am reading into getting started on my small-scale DevOps. Been trying out GitHub workflows and codespaces, stull fumbling around but I will get there. Any further pointers would be greatly Appreciated

**Reply by ayodejiayodele** [reply]
That's great @AngelofVerdant, I'm glad this helped. Here are some good pointers to get you started, I particularly love the skills one:

1. [GitHub Skills](https://github.com/skills/)
2. [GitHub Resources](https://resources.github.com/)

**Reply by mahmood521** [reply]
سلام فارسی بنویسم مشکلی نداره

**Comment by Fumrizh** [comment]
hard make you a breath

**Comment by clins10** [comment]
this is really helpful. thanks @ayodejiayodele

**Reply by ayodejiayodele** [reply]
I'm glad you enjoyed reading it. Thanks @clins10 and have a great 2023 ahead 🍾🎉

**Reply by Oyebobo** [reply]
Hi @ayodejiayodele, 

May I introduce myself please? My name is Oyebola, and I wondered if I can buddy up with you as a mentee under your mentorship. I know this sounds all so abrupt, but I will be happy to provide you a run through on my background if you will let me. Hopefully, it will be modest and humble enough for you to take me up as a community mentee.

**Reply by horizonvert1027** [reply]
May I introduce myself?
My name is Ainas, I am a professional web (React.js/Vue) and mobile developer with 5+ years of experience. I prefer to code from scratch and create fully custom websites. My designs follow a minimalist approach and focus on simplistic powerful aesthetics. I have completed a wide range of projects for 5+ years.

I love learning new frameworks and design styles and push myself to stay up to date with all new web technologies.

What I'm perfect at developing applications fast, in short terms, according to client requirements with Pixel Perfect quality! I consider myself creative and hardworking, as my stats show. If you liked my profile so far, drop me a message, and we'll discuss how I can be useful to your project.

My core skills:
- ReactJS/Redux, Typescript
- React Native, Ionic, Cordova
- Node.js/Express, Python
- PHP, Laravel, CMS
- Bootstrap, Material UI, Kendo UI
- Typescript, Javascript/es6+
- Firebase, Aws
- Scss, Sass, Less, CSS3
- Github, Bitbucket, Gitlab
- Trello, Jira, Asana

**Comment by abodaa198911** [comment]
10000

**Comment by Vergilraven** [comment]
这是来自QQ邮箱的假期自动回复邮件。你好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。

**Comment by ilhom4177** [comment]
salom

**Comment by mahmood521** [comment]
سلام و عرض ادب و احترام به همه ی دوستان 
خواستم بگم من تازه کار هستم لطفا تو پروژه هوای منو داشته باشید تشکر

**Comment by aboodic** [comment]
Aboodic

**Comment by bzkarim250** [comment]
Thanks for sharing!

**Comment by NikolaRoganovic** [comment]
Very good text, GJ !

**Comment by horizonvert1027** [comment]
I appreciate everyone's effort.
We shall achieve good relations and results as each other's effort.

**Comment by nkuja7** [comment]
On Wed, 25 Jan 2023 at 11:59 PM, Oyebola ***@***.***> wrote:

> Hi @ayodejiayodele <https://github.com/ayodejiayodele>,
>
> May I introduce myself please? My name is Oyebola, and I wondered if I can
> buddy up with you as a mentee under your mentorship. I know this sounds all
> so abrupt, but I will be happy to provide you a run through on my
> background if you will let me. Hopefully, it will be modest and humble
> enough for you to take me up as a community mentee.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/community/discussions/41489#discussioncomment-4782222>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5KRIHCXKEGYUYKERO5SDATWUGO4RANCNFSM6AAAAAAS3YJB7E>
> .
> You are receiving this because you are subscribed to this thread. Message
> ID: ***@***.***
> com>
>
    Hi hope you are well too, And also my name is
    Nkululeko by the way so yeah I’m also up for       that? Let’s do hard!

**Reply by white10146** [reply]
I needed help sir on how to add my token to my trust wallet

**Comment by nkuja7** [comment]
Hi hope you are well, my name is Nkululeko and
   I’d also love to be part of the team so yeah?
   Let’s do hard!

**Comment by Oyebobo** [comment]
Thank you for introducing yourself.
I am very pleased by your introduction, and prospect of using your help in my very modest project.
My best coding achievement till date is a personal art-ehibition webpage which I struggle to edit with Visual Studio Code.
I am essentially looking for a mentor who can help me refine my skills and navigate specialist coding tools like Visual Studio.
If you can help, I guess the next question is, how can we begin a mentor-mentee relationship?

With thanks
Alfred Oyebola Ogun
________________________________

**Comment by nkuja7** [comment]
I think it would much better if we first make use of Slack that would be a
much wise idea if we utilised that platform.

**Comment by MattewAdams** [comment]
Guys, can I join here either? I see you need help with this. For several years now, I've been using different ways for meeting, but recently I decided to find something like conferences, so I turned to [Digital Samba API](https://www.digitalsamba.com/video-api). I'm good at it. I think it's interesting to find out what development is.

**Reply by horizonvert1027** [reply]
thanks for your reply.
let me check.

**Comment by Aacayipbaris87** [comment]
hello can you contact me bariskarabeyin@gmail.com

**Comment by Oyebobo** [comment]
Hi Matthews Adams,

I have just signed up to Digital Samba API.
Is there anyway we can connect in a session room.
I hope to use your mentorship to improve my development skills.

With thanks
Oyebola Alfred

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Maximiliano17** [comment]
nice

**Reply by JuampaVLB** [reply]
?

**Comment by danielapassos** [comment]
This is a great thread!

3 more to the list 👇 
> 1. Insufficient automation and reliance on manual processes.
> 2. Lack of standardization and consistency across teams and environments.
> 3. Poorly defined roles and responsibilities, leading to confusion and duplication of effort.

**Comment by Malory229** [comment]
I am new on GitHub and I wish to learn more about how to use it. Thank you for sharing about DevOps. I was interested to learn about it.

**Comment by hakimexpo** [comment]
https://www.facebook.com/profile.php?id=100072236438292

**Comment by hakimexpo** [comment]
more

**Comment by hakimexpo** [comment]
https://www.facebook.com/profile.php?id=100072236438292

**Comment by umair22556** [comment]
Who are you

**Comment by umair22556** [comment]
Yes how I can help you with

**Comment by ErlinLopez17** [comment]
I am new on GitHub and I wish to learn more about how to use it. Thank you for sharing about DevOps. I was interested to learn about it.

**Comment by goran-revops** [comment]
By encouraging small changes and iteration over time, teams can gradually improve their processes and achieve better results. It's an exciting and challenging area to work in, but the rewards are well worth the effort.

**Comment by karthik1896** [comment]
Good Thread for the people's who are practicing DevOps and DevOps is a practice which needs to be inherited in every organizations who are playing in Cloud Platforms. Also using Docker , Kubernetes and Jenkins organization can deploy their application in any environment with faster deployments

**Comment by Genniefrmbx** [comment]
@LCNMNGMNT
